### PR TITLE
BUG: interchange protocol with nullable datatypes a non-null validity

### LINKED ISSUE #57643: BUG: interchange protocol with nullable datatypes a non-null validity provides nonsense results
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pyarrow.interchange as pai
import pandas as pd
import polars as pl

data = pd.DataFrame(
    {
        "b1": pd.Series([1, 2, None], dtype='Int64'),
    }
)
print(pai.from_dataframe(data))
print(pl.from_dataframe(data))
```

### Issue Description

Result:
```
pyarrow.Table
b1: int64
----
b1: [[4607182418800017408,4611686018427387904,9221120237041090560]]
shape: (3, 1)
┌─────────────────────┐
│ b1                  │
│ ---                 │
│ i64                 │
╞═════════════════════╡
│ 4607182418800017408 │
│ 4611686018427387904 │
│ 9221120237041090560 │
└─────────────────────┘
```

### Expected Behavior

`[1, 2, None]`

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit                : 92a52e231534de236c4e878008a4365b4b1da291
python                : 3.10.12.final.0
python-bits           : 64
OS                    : Linux
OS-release            : 5.15.133.1-microsoft-standard-WSL2
Version               : #1 SMP Thu Oct 5 21:02:42 UTC 2023
machine               : x86_64
processor             : x86_64
byteorder             : little
LC_ALL                : None
LANG                  : C.UTF-8
LOCALE                : en_US.UTF-8

pandas                : 3.0.0.dev0+355.g92a52e2315
numpy                 : 1.26.4
pytz                  : 2024.1
dateutil              : 2.8.2
setuptools            : 59.6.0
pip                   : 24.0
Cython                : 3.0.8
pytest                : 8.0.0
hypothesis            : 6.98.6
sphinx                : 7.2.6
blosc                 : None
feather               : None
xlsxwriter            : 3.1.9
lxml.etree            : 5.1.0
html5lib              : 1.1
pymysql               : 1.4.6
psycopg2              : 2.9.9
jinja2                : 3.1.3
IPython               : 8.21.0
pandas_datareader     : None
adbc-driver-postgresql: None
adbc-driver-sqlite    : None
bs4                   : 4.12.3
bottleneck            : 1.3.7
fastparquet           : 2024.2.0
fsspec                : 2024.2.0
gcsfs                 : 2024.2.0
matplotlib            : 3.8.3
numba                 : 0.59.0
numexpr               : 2.9.0
odfpy                 : None
openpyxl              : 3.1.2
pyarrow               : 15.0.0
pyreadstat            : 1.2.6
python-calamine       : None
pyxlsb                : 1.0.10
s3fs                  : 2024.2.0
scipy                 : 1.12.0
sqlalchemy            : 2.0.27
tables                : 3.9.2
tabulate              : 0.9.0
xarray                : 2024.1.1
xlrd                  : 2.0.1
zstandard             : 0.22.0
tzdata                : 2024.1
qtpy                  : 2.4.1
pyqt5                 : None

</details>

### PR DESCRIPTION
**Pull request by MarcoGorelli**
- [ ] closes #57643 (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

This deals with pandas nullable dtypes

pyarrow nullables ones will require changes to the PandasBuffer class, so I'll handle those separately

### CODE REVIEW

**Review by WillAyd — APPROVED** [review]

File: pandas/tests/interchange/test_impl.py [hunk 1]
```diff
@@ -417,17 +416,35 @@ def test_non_str_names_w_duplicates():
         pd.api.interchange.from_dataframe(dfi, allow_copy=False)
 
 
-@pytest.mark.parametrize(
-    "dtype", ["Int8", pytest.param("Int8[pyarrow]", marks=td.skip_if_no("pyarrow"))]
-)
-def test_nullable_integers(dtype: str) -> None:
+def test_nullable_integers() -> None:
+    # https://github.com/pandas-dev/pandas/issues/55069
+    df = pd.DataFrame({"a": [1]}, dtype="Int8")
```
Comment (mroeschke) [inline pandas/tests/interchange/test_impl.py:421]: Could you include a `None` in here to test null?

And parametrize over `["Int8", "Float32", "boolean"]`?

File: pandas/core/interchange/column.py [hunk 2]
```diff
@@ -190,6 +190,10 @@ def describe_categorical(self):
 
     @property
     def describe_null(self):
+        if isinstance(self._col.dtype, BaseMaskedDtype):
+            column_null_dtype = ColumnNullType.USE_BYTEMASK
+            null_value = 1
```
Comment (mroeschke) [inline pandas/core/interchange/column.py:195]: Curious why 1 here. Is this a throwaway value?

File: pandas/tests/interchange/test_impl.py [hunk 3]
```diff
@@ -417,17 +416,35 @@ def test_non_str_names_w_duplicates():
...
+    df = pd.DataFrame({"a": [1]}, dtype="Int8[pyarrow]")
     expected = pd.DataFrame({"a": [1]}, dtype="int8")
     result = pd.api.interchange.from_dataframe(df.__dataframe__())
     tm.assert_frame_equal(result, expected)
 
 
+def test_nullable_integers_w_missing_values() -> None:
+    # https://github.com/pandas-dev/pandas/issues/57643
+    pytest.importorskip("pyarrow", "11.0.0")
+    import pyarrow.interchange as pai
```
Comment (mroeschke) [inline pandas/tests/interchange/test_impl.py:439]:
```suggestion
    pai = pytest.importorskip("pyarrow.interchange", "11.0.0")
```

File: pandas/core/interchange/column.py (hunk 2 above)
Comment (WillAyd) [reply pandas/core/interchange/column.py:195]: 1 = true for the uint8_t byte mask we would get back from NumPy in a BaseMaskedDtype

Comment (MarcoGorelli) [reply pandas/core/interchange/column.py:195]: `describe_null` returns a tuple:

https://github.com/pandas-dev/pandas/blob/4019d41816e6114d41731638fe027a8b0597db72/pandas/core/interchange/dataframe_protocol.py#L291-L301

In this case, we have a bytemask, where `1` indicates missing values

File: pandas/tests/interchange/test_impl.py (hunk 3 above)
Comment (MarcoGorelli) [reply pandas/tests/interchange/test_impl.py:439]: this doesn't work for me, it just skips the test entirely if I do that

File: pandas/tests/interchange/test_impl.py (hunk 1 above)
Comment (MarcoGorelli) [reply pandas/tests/interchange/test_impl.py:421]: boolean isn't supported yet, there's a separate issue about that: https://github.com/pandas-dev/pandas/issues/55332

sure, I've added a separate test `test_pandas_nullable_w_missing_values` which includes a null

**Review by mroeschke — APPROVED** [review]

File: doc/source/whatsnew/v2.2.1.rst [hunk 4]
```diff
@@ -19,6 +19,7 @@ Enhancements
 
 Fixed regressions
 ~~~~~~~~~~~~~~~~~
+- :meth:`DataFrame.__dataframe__` was producing incorrect data buffers when the a column's type was a pandas nullable on with missing values (:issue:`56702`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v2.2.1.rst:22]: Oh actually. I guess we need a 2.2.2 whatsnew first

**Review by mroeschke — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v2.2.1.rst (hunk 4 above)
Comment (MarcoGorelli) [reply doc/source/whatsnew/v2.2.1.rst:22]: ah right, 2.2.1 is already out, sorry about that, thanks for spotting it

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @MarcoGorelli

**Pull request merged**